File: examples/java/02-variables-and-types/Main.java
/**
 * Variables and Types in Java
 * ===========================
 * Java uses static, strong, nominal typing.
 * Primitive types live on stack; reference types on heap.
 */
public class Main {
    public static void main(String[] args) {

        // ── Primitive types ──────────────────────────────

        byte   tiny     = 127;          // 8-bit,  [-128, 127]
        short  small    = 32_767;       // 16-bit
        int    count    = 42;           // 32-bit
        long   large    = 9_000_000_000_000_000L;  // 64-bit
        float  approx   = 3.14f;       // 32-bit IEEE 754
        double pi       = 3.14159;     // 64-bit IEEE 754
        boolean active  = true;
        char   letter   = 'A';         // 16-bit Unicode (UTF-16)

        System.out.println("count = " + count + " (int)");
        System.out.println("pi = " + pi + " (double)");
        System.out.println("active = " + active + " (boolean)");
        System.out.println("letter = " + letter + " (char, value " + (int) letter + ")");

        // ── Widening vs narrowing ────────────────────────

        System.out.println("\n--- Widening vs narrowing ---");

        // Widening: smaller → larger (implicit, safe)
        int    a = 42;
        double b = a;   // int → double, automatic
        System.out.println("int " + a + " → double " + b + " (widening, implicit)");

        // Narrowing: larger → smaller (requires cast, may lose data)
        double d = 3.99;
        int    i = (int) d;  // truncates to 3!
        System.out.println("double " + d + " → int " + i + " (narrowing, explicit cast)");

        // String to number — explicit parsing required
        int parsed = Integer.parseInt("42");
        System.out.println("Integer.parseInt(\"42\") = " + parsed);

        // "3" + 4 → "34" (string concatenation, not addition!)
        String surprise = "3" + 4;
        System.out.println("\"3\" + 4 = \"" + surprise + "\" (concatenation!)");

        // ── Reference types ──────────────────────────────

        System.out.println("\n--- Reference types ---");

        String name = "Software Atlas";
        System.out.println("name = " + name);
        System.out.println("name.length() = " + name.length());
        System.out.println("name.charAt(0) = " + name.charAt(0));

        // Strings are immutable
        String upper = name.toUpperCase();  // returns NEW string
        System.out.println("name = " + name + " (unchanged)");
        System.out.println("upper = " + upper);

        // ── Autoboxing (primitive ↔ wrapper) ─────────────

        System.out.println("\n--- Autoboxing ---");

        Integer boxed = count;        // int → Integer (autoboxing)
        int unboxed = boxed;          // Integer → int (auto-unboxing)
        System.out.println("boxed = " + boxed + " (Integer)");
        System.out.println("unboxed = " + unboxed + " (int)");

        // Careful: Integer can be null, int cannot!
        Integer nullable = null;
        // int crash = nullable;  // NullPointerException at runtime!

        // ── var (Java 10+) ───────────────────────────────

        System.out.println("\n--- var (local type inference) ---");

        var message = "Hello, Atlas!";  // inferred as String
        var number = 42;                // inferred as int
        var list = java.util.List.of("a", "b", "c");  // inferred as List<String>

        System.out.println("message: " + message + " (" + message.getClass().getSimpleName() + ")");
        System.out.println("list: " + list + " (" + list.getClass().getSimpleName() + ")");

        // var cannot be used for fields, parameters, or return types

        // ── Arrays ───────────────────────────────────────

        System.out.println("\n--- Arrays ---");

        int[] arr = {10, 20, 30, 40, 50};
        System.out.println("arr.length = " + arr.length);
        System.out.println("arr[0] = " + arr[0]);

        // Bounds checking! (unlike C)
        try {
            int oob = arr[10];
        } catch (ArrayIndexOutOfBoundsException e) {
            System.out.println("arr[10] → " + e.getClass().getSimpleName());
        }

        // ── Collections ──────────────────────────────────

        System.out.println("\n--- Collections ---");

        // List (mutable)
        var fruits = new java.util.ArrayList<>(java.util.List.of("apple", "banana"));
        fruits.add("cherry");
        System.out.println("fruits = " + fruits);

        // Map
        var ages = java.util.Map.of("Ada", 36, "Grace", 85);
        System.out.println("ages = " + ages);

        // Set
        var tags = java.util.Set.of("java", "types", "static");
        System.out.println("tags = " + tags);

        // ── Records (Java 16+) ───────────────────────────

        System.out.println("\n--- Records (Java 16+) ---");

        record Point(double x, double y) {
            double distanceFromOrigin() {
                return Math.sqrt(x * x + y * y);
            }
        }

        var p = new Point(3.0, 4.0);
        System.out.println("point = " + p);
        System.out.println("distance = " + p.distanceFromOrigin());
        System.out.println("x = " + p.x() + ", y = " + p.y());

        // Records are immutable!
        // p.x = 5.0;  // COMPILE ERROR

        // ── Null — billion-dollar mistake ────────────

        System.out.println("\n--- Null ---");

        String s = null;
        System.out.println("s = " + s);  // prints "null"
        try {
            int len = s.length();  // NullPointerException!
        } catch (NullPointerException e) {
            System.out.println("s.length() → NullPointerException");
        }

        // Optional (Java 8+) — explicit "maybe a value"
        var opt = java.util.Optional.ofNullable(s);
        String result = opt.orElse("(default)");
        System.out.println("Optional: " + result);

        // ── Nominal typing ───────────────────────────────

        System.out.println("\n--- Nominal typing ---");

        // Java uses nominal typing: types must explicitly declare relationships
        // Even if two classes have identical methods, they're not interchangeable

        interface Quacker {
            String quack();
        }

        class Duck implements Quacker {
            public String quack() { return "Quack!"; }
        }

        class Person implements Quacker {
            public String quack() { return "I'm quacking!"; }
        }

        // Both must explicitly "implements Quacker"
        Quacker q = new Duck();
        System.out.println("Duck: " + q.quack());
        q = new Person();
        System.out.println("Person: " + q.quack());

        // ── Summary ──────────────────────────────────────

        System.out.println("\n--- Summary ---");
        System.out.println("Java: static + strong + nominal typing");
        System.out.println("  - Types checked at compile time (static)");
        System.out.println("  - No implicit narrowing (strong)");
        System.out.println("  - Must declare 'implements' (nominal)");
        System.out.println("  - Primitive types (stack) + reference types (heap)");
        System.out.println("  - Null exists (billion-dollar mistake)");
        System.out.println("  - Records for immutable data (16+)");
        System.out.println("  - var for local type inference (10+)");
    }
}

File: examples/go/02-variables-and-types/main.go
package main

import (
	"fmt"
	"math"
	"reflect"
)

// ── Types for structural typing demo ─────────────────────

// Quacker is an interface — any type with Quack() string satisfies it.
// No "implements" keyword needed.
type Quacker interface {
	Quack() string
}

type Duck struct{}

func (d Duck) Quack() string { return "Quack!" }

type Person struct{ Name string }

func (p Person) Quack() string { return p.Name + " quacks!" }

// ── Point struct ─────────────────────────────────────────

type Point struct {
	X, Y float64
}

func (p Point) DistanceFromOrigin() float64 {
	return math.Sqrt(p.X*p.X + p.Y*p.Y)
}

// ── Main ─────────────────────────────────────────────────

func main() {
	// ── Basic types ──────────────────────────────────────

	// Explicit type declaration
	var count int = 42
	fmt.Printf("count = %d, type = %T\n", count, count)

	// Type inference with :=
	pi := 3.14159 // inferred as float64
	fmt.Printf("pi = %f, type = %T\n", pi, pi)

	active := true // inferred as bool
	fmt.Printf("active = %t, type = %T\n", active, active)

	name := "Software Engineering Atlas" // inferred as string
	fmt.Printf("name = %s, type = %T\n", name, name)

	// ── Zero values ──────────────────────────────────────
	// Every type has a well-defined zero value. No uninitialised variables.

	fmt.Println("\n--- Zero values ---")
	var zeroInt int
	var zeroFloat float64
	var zeroBool bool
	var zeroString string
	var zeroPointer *int

	fmt.Printf("int:     %d\n", zeroInt)    // 0
	fmt.Printf("float64: %f\n", zeroFloat)  // 0.0
	fmt.Printf("bool:    %t\n", zeroBool)     // false
	fmt.Printf("string:  %q\n", zeroString)   // "" (empty)
	fmt.Printf("pointer: %v\n", zeroPointer) // <nil>

	// ── Integer types ────────────────────────────────────
	// Go has explicit sized integers: int8, int16, int32, int64
	// and unsigned variants: uint8, uint16, uint32, uint64
	// `int` is platform-dependent (32 or 64 bit)

	fmt.Println("\n--- Integer types ---")
	var small int8 = 127 // max value for int8
	var medium int32 = 2_000_000_000
	var large int64 = 9_000_000_000_000_000_000
	var unsigned uint8 = 255 // byte alias

	fmt.Printf("int8:   %d (max %d)\n", small, math.MaxInt8)
	fmt.Printf("int32:  %d\n", medium)
	fmt.Printf("int64:  %d\n", large)
	fmt.Printf("uint8:  %d (max %d)\n", unsigned, math.MaxUint8)

	// ── No implicit conversions ──────────────────────────
	// Go does NOT implicitly convert between types, even numeric ones

	fmt.Println("\n--- No implicit conversions ---")
	var a int32 = 10
	var b int64 = 20

	// c := a + b  // COMPILE ERROR: mismatched types int32 and int64

	c := int64(a) + b // explicit conversion required
	fmt.Printf("int64(%d) + %d = %d\n", a, b, c)

	// ── Strings ──────────────────────────────────────────

	fmt.Println("\n--- Strings ---")
	greeting := "Hello, 世界" // UTF-8 by default
	fmt.Printf("greeting = %s\n", greeting)
	fmt.Printf("len (bytes) = %d\n", len(greeting))

	// Iterate over runes (Unicode code points), not bytes
	for i, r := range greeting {
		fmt.Printf("  [%d] %c (U+%04X)\n", i, r, r)
	}

	// Strings are immutable
	// greeting[0] = 'h'  // COMPILE ERROR

	// ── Arrays and slices ────────────────────────────────

	fmt.Println("\n--- Arrays and slices ---")

	// Array — fixed size, part of the type
	arr := [3]int{10, 20, 30}
	fmt.Printf("array = %v, type = %T\n", arr, arr)

	// Slice — dynamic, reference to underlying array
	fruits := []string{"apple", "banana", "cherry"}
	fruits = append(fruits, "date")
	fmt.Printf("slice = %v, len=%d, cap=%d\n", fruits, len(fruits), cap(fruits))

	// ── Maps ─────────────────────────────────────────────

	fmt.Println("\n--- Maps ---")
	ages := map[string]int{
		"Ada":   36,
		"Grace": 85,
	}
	fmt.Printf("ages = %v\n", ages)

	// Accessing a missing key returns zero value
	score, ok := ages["Alan"]
	fmt.Printf("ages[\"Alan\"] = %d, ok = %t\n", score, ok)

	// ── Structs ──────────────────────────────────────────

	fmt.Println("\n--- Structs ---")
	p := Point{X: 3.0, Y: 4.0}
	fmt.Printf("point = %+v\n", p)
	fmt.Printf("distance from origin = %.2f\n", p.DistanceFromOrigin())

	// ── Pointers ─────────────────────────────────────────

	fmt.Println("\n--- Pointers ---")
	x := 42
	ptr := &x // pointer to x
	fmt.Printf("x = %d, ptr = %p, *ptr = %d\n", x, ptr, *ptr)

	*ptr = 100 // modify x through pointer
	fmt.Printf("after *ptr = 100: x = %d\n", x)
	// No pointer arithmetic! (unlike C)

	// ── Structural typing (interfaces) ───────────────────

	fmt.Println("\n--- Structural typing ---")

	// Both Duck and Person satisfy Quacker interface
	// because they have a Quack() string method.
	// No "implements" keyword — it's automatic.

	var q Quacker

	q = Duck{}
	fmt.Printf("Duck: %s\n", q.Quack())

	q = Person{Name: "Bob"}
	fmt.Printf("Person: %s\n", q.Quack())

	// ── Type reflection ──────────────────────────────────

	fmt.Println("\n--- Reflection ---")
	values := []any{42, 3.14, "hello", true}
	for _, v := range values {
		fmt.Printf("  value = %-8v type = %v\n", v, reflect.TypeOf(v))
	}

	// ── Summary ──────────────────────────────────────────

	fmt.Println("\n--- Summary ---")
	fmt.Println("Go: static + strong typing")
	fmt.Println("  - Types checked at compile time (static)")
	fmt.Println("  - No implicit conversions (strong)")
	fmt.Println("  - Type inference with :=")
	fmt.Println("  - Zero values — no uninitialised variables")
	fmt.Println("  - Structural typing for interfaces")
	fmt.Println("  - Sized integers (int8..int64)")
}
